Scripts for 2ndline and republishing tasks

tuzz opened the pull request:
I've finished my manual testing and this is now open for review.

I did see a few peculiarites around republishing, but I think these are out-of-scope for this ticket as they are captured in this one: https://trello.com/c/IAknDPK2/9-data-comparison-of-raib-between-v1-and-v2

I will add "republishing should work" as an A/C of the data comparison / fixes story.

There's a pull request to update the opsmanual here: https://github.gds/gds/opsmanual/pull/739

tuzz committed:
Add a Republisher for queueing republish sidekiq jobs

tuzz committed:
Add a RepublishWorker that republishes content

tuzz committed:
Add ‘previous_version’ to a document’s attributes…

This is an opt-in feature of the Publishing API.
See: https://github.com/alphagov/publishing-api/blob/master/doc/publishing-api-syntactic-usage.md

tuzz committed:
Add a Document#discard method

For now, this will only be used by a rake task but
we could add this to the UI fairly easily in the
future.

tuzz committed:
Add a rake task for discarding a draft

h-lame commented on app/models/document.rb:
```diff
@@ -256,6 +256,14 @@ def withdraw
+
+  def discard!
+    Services.publishing_api.discard_draft(content_id, previous_version: previous_version)
+  end
```
This pair of methods has different semantics to the other publishing-api write methods.  `save`, `publish!`, and `unpublish` all use `handle_remote_error` to wrap the API calls and return `true` or `false` rather than raise.  It's not clear why `publish!` has a bang when the others don't.  

I'm not sure we're following the `ActiveRecord` pattern of `!` meaning a method that might raise and the unadorned method meaning one that won't.  It's a familiar pattern though so maybe we should?  On the other hand do we even need the raise variants, we're not currently using them and it's not clear we'd want to?

Either way, this new pair of methods stands out as odd compared to what we have and we should choose how to make discarding feel more similar to the other write methods.

tuzz replied:
I don't think we need raise variants of those methods yet, so I'll remove `discard!`. We might want them eventually, but simpler is better for now. 👍 

I think I'll rename `publish!` to `publish` too.

tuzz committed:
Remove the ‘!’ variant of the discard method

tuzz committed:
Rename #publish! to #publish

This method doesn’t raise an error if it fails and
so I wouldn’t expect it to be suffixed with a ‘!’

tuzz committed:
Do not republish things that can cause side-effects…

If we were to republish something in an
‘unpublished’ state, it exposes a risk that it
could be made accidentally live if a subsequent
unpublished request fails. There is no way to
ensure that these operations happen as a single
atomic operation in the Publishing API.

Our best option in this case is to skip the
republishing of this document. It may be the case
that whoever runs the republishing script needs
to do something to account for those documents
that have been skipped.

The RepublishWorker will skip over any other
states that it doesn’t know about (or that may be
introduced in the future) as it doesn’t know the
correct set of requests to issue to ensure the
document is republished into that state.

tuzz committed:
Update gds-api-adapters to 31.1.0

tuzz committed:
Add the remaining ops tasks

tuzz committed:
Add a task to grant user permissions

tuzz committed:
Use ‘pry-rails’ so that rails console uses pry

tuzz committed:
Make `rake republish:one` run synchronously

This reduces the number of moving parts and should
be easier to debug.

tuzz committed:
Fix some tests broken during rebase

christophwong commented on README.md:
````diff
@@ -40,9 +40,24 @@ $ ./startup.sh
+You also need to set the `app_name` for the Dummy API User in Publishing API:
+
+```ruby
+User.find_by(email: "dummyapiuser@domain.com").update!(app_name: "specialist-publisher")
````
Can we consolidate this blocks of text with line 65 - 76? https://github.com/alphagov/specialist-publisher-rebuild/pull/802/files#diff-04c6e90faac2675aa89e2176d2eec7d8R74

tuzz replied:
Good catch. I've removed that content. I think this covers all of it and it's a bit shorter. What do you think?

christophwong commented on spec/workers/republish_worker_spec.rb:
```diff
@@ -0,0 +1,101 @@
+        subject.perform(content_id)
+
+        expect(WebMock).not_to have_requested(:post, /search/)
+      end
```
would it be worthwhile to add "does not speak to email alert api" as a test as well?

tuzz replied:
added 👍

tuzz committed:
Ensure that republishing does not trigger emails

tuzz committed:
Remove outdated documentation
